README.md
#  PodlovePlugin plugin
The PodlovePlugin plugin is an audio player plugin for [Urlaub.be](https://github.com/urlaube/urlaube) using the [Podlove Web Player](https://github.com/podlove/podlove-ui).

## Installation
Place the folder containing the plugin into your plugins directory located at `./user/plugins/`.

## Configuration
To configure the plugin you can change the corresponding settings in your configuration file located at `./user/config/config.php`.

### Subscribe Button
The subscribe button is shown as soon as a feed is set.
```php
Plugins::set("PODLOVE_FEED_URL", null);
```

A few services require additional information besides a feed:

```php
// https://podcasts.apple.com/podcast/[id]
Plugins::set("PODLOVE_APPLE_PODCAST_ID", null);

// https://www.deezer.com/en/show/[id]
Plugins::set("PODLOVE_DEEZER_ID", null);

// https://soundcloud.com/[id]
Plugins::set("PODLOVE_SOUNDCLOUD_ID", null);

// https://open.spotify.com/show/[id]
Plugins::set("PODLOVE_SPOTIFY_ID", null);

// https://www.stitcher.com/podcast/[id]
Plugins::set("PODLOVE_STITCHER_ID", null);

// https://www.youtube.com/channel/[id]
Plugins::set("PODLOVE_YOUTUBE_ID", null);

// Castbox
Plugins::set("PODLOVE_CASTBOX_ID", null);
```

### Player Settings

```php
// One of the follwing: [chapters, files, share, playlist]
Plugins::set("PODLOVE_ACTIVE_TAB", null);

// To disable embedding set it to null
Plugins::set("PODLOVE_EMBEDDING", "/share.html");

// Interactive theming engine: https://docs.podlove.org/podlove-web-player/theme.html
Plugins::set("PODLOVE_THEME_COLORS", [
        "brand" => "#166255",
        "brandDark" => "#166255",
        "brandDarkest" => "#1A3A4A",
        "brandLightest" => "#E5EAECFF",
        "shadeDark" => "#807E7C",
        "shadeBase" => "#807E7C",
        "contrast" => "#000",
        "alt" => "#fff"
      ]);

// Show share links for ["facebook","twitter","whats-app","linkedin","pinterest","xing","mail","link"]
Plugins::set("PODLOVE_SHARE_CHANNELS", ["facebook", "twitter", "mail", "link", "whats-app"]);

// Allow to add current play position to shared links 
Plugins::set("PODLOVE_SHARE_PLAYTIME", true);
```

## Usage

To include the audio player on a page you have to use the `[podloveplayer]` shortcode within the content. It will be replaced with the HTML sourcecode of the audio player.

To use the plugin you can have to values to the header of your content files located at `/user/content/*`.

### Required
You can add multiple audio files with diffrent codecs (seperated by `,`):
```
podlove_audioFiles:
podlove_audioSizes:
podlove_audioTitles:
podlove_audioMimeTypes:
podlove_duration:
```

### Optional
The following tags are optional but improve the user experience.

Additional information about the audiofile. Chapters are seperated by `,`.

```
podlove_title:
podlove_poster:
podlove_duration:

podlove_chapterTitles:
podlove_chapterStarts:
podlove_chapterLinks:
podlove_chapterImages:
```



Files to download connected to the audio file (seperated by `,`). Audio files for the player are automatically added.

```
podlove_downloadFiles:
podlove_downloadSizes:
podlove_downloadTitles:
podlove_downloadMimeTypes:
```

CHANGELOG.md
# Changelog

## 0.1a0 (10.04.2020)
### Features
* initial podcastaudioplugin version

## 0.2 (27.05.2020)
### Features
* initial podlove version

lib/variant-l.js
// [AIV_SHORT]  Build version: 5.1.6 - Tuesday, May 26th, 2020, 6:13:15 PM  
 (window.webpackJsonp=window.webpackJsonp||[]).push([["variant-l"],{"./src/templates/variant-l.html":function(e,t){e.exports='<root data-test=player--l style=max-width:950px;min-width:260px> <div class="p-4 flex flex-col"> <div class="flex-col items-center mobile:flex tablet:hidden"> <show-title class=text-sm></show-title> <episode-title class="text-base mb-2"></episode-title> <subscribe-button class="mb-4 mobile:flex tablet:hidden"></subscribe-button> <poster class="rounded-sm w-48 shadow overflow-hidden"></poster> <divider class="w-full my-4"></divider> </div> <div class="tablet:flex flex-grow"> <div class="w-40 mobile:hidden tablet:block tablet:mr-6"> <poster class="rounded-sm shadow overflow-hidden"></poster> </div> <div class=w-full> <div class="hidden tablet:block"> <div class="flex justify-between items-center"> <div> <show-title class=text-base></show-title> <episode-title class=text-xl></episode-title> </div> <subscribe-button></subscribe-button> </div> <divider class="w-full my-4"></divider> </div> <div class="flex w-full items-center justify-center"> <play-button class="block mr-5" variant=simple></play-button> <timer-current class="text-base mr-5 hidden tablet:block"></timer-current> <progress-bar class=mr-5></progress-bar> <timer-duration class=text-base></timer-duration> </div> </div> </div> </div> <error></error> </root> '}}]); 

lib/variant-m.js
// [AIV_SHORT]  Build version: 5.1.6 - Tuesday, May 26th, 2020, 6:13:15 PM  
 (window.webpackJsonp=window.webpackJsonp||[]).push([["variant-m"],{"./src/templates/variant-m.html":function(t,s){t.exports='<root data-test=player--m style=max-width:950px;min-width:260px> <div class="p-4 flex flex-col"> <div class="flex flex-grow mb-3"> <div class="w-20 mr-4"> <poster class="rounded-sm shadow overflow-hidden"></poster> </div> <div class="w-full tablet:my-2"> <show-title class=text-sm></show-title> <episode-title class=text-base></episode-title> </div> </div> <div class="flex items-center justify-center"> <subscribe-button></subscribe-button> </div> <divider class="w-full my-4"></divider> <div class="flex w-full items-center justify-center"> <play-button class="block mr-5" variant=simple></play-button> <timer-current class="text-sm mr-5 hidden tablet:block"></timer-current> <progress-bar class=mr-5></progress-bar> <timer-duration class=text-sm></timer-duration> </div> </div> <error></error> </root> '}}]); 

lib/variant-xl.js
// [AIV_SHORT]  Build version: 5.1.6 - Tuesday, May 26th, 2020, 6:13:15 PM  
 (window.webpackJsonp=window.webpackJsonp||[]).push([["variant-xl"],{"./src/templates/variant-xl.html":function(t,e){t.exports='<root data-test=player--xl style=max-width:1200px;min-width:260px> <div class="tablet:px-6 tablet:pt-6 mobile:px-4 mobile:pt-4 flex flex-col"> <div class="flex-col items-center mobile:flex tablet:hidden"> <show-title class=text-sm></show-title> <episode-title class="text-base mb-2"></episode-title> <subscribe-button class="mb-4 mobile:flex tablet:hidden"></subscribe-button> <poster class="rounded-sm w-48 shadow overflow-hidden"></poster> <divider class="w-full my-6"></divider> </div> <div class="tablet:flex flex-grow"> <div class="w-64 mobile:hidden tablet:block tablet:mr-6"> <poster class="rounded-sm shadow overflow-hidden"></poster> </div> <div class=w-full> <div class="hidden tablet:block"> <show-title class=text-base></show-title> <episode-title class="text-xl desktop:text-2xl"></episode-title> <divider class="w-full my-4"></divider> </div> <div class="flex items-center justify-between"> <div class=block> <play-state on=active> <speed-control class="flex items-center"></speed-control> </play-state> </div> <div class=flex> <play-state on=active> <chapter-previous class="mx-2 block"></chapter-previous> </play-state> <play-state on=active> <step-backward class="mx-2 block"></step-backward> </play-state> <play-button class="mx-2 block" :label="$t(\'PLAYER.PLAY_EPISODE\')"></play-button> <play-state on=active> <step-forward class="mx-2 block"></step-forward> </play-state> <play-state on=active> <chapter-next class="mx-2 block"></chapter-next> </play-state> </div> <div class=block> <play-state on=active> <volume-control class="flex items-center"></volume-control> </play-state> </div> </div> <div class="flex w-full"> <progress-bar></progress-bar> </div> <div class="flex w-full -mt-2"> <div class="w-3/12 text-left"> <timer-current class=text-sm></timer-current> </div> <div class="w-6/12 text-center truncate"> <play-state on=active> <current-chapter class=text-sm></current-chapter> </play-state> </div> <div class="w-3/12 text-right"> <timer-duration class=text-sm></timer-duration> </div> </div> </div> </div> <divider class="w-full mt-6 mb-3"></divider> <div class="flex justify-between"> <div class="flex mobile:w-full tablet:w-3/12 desktop:w-3/12 justify-between"> <tab-trigger tab=chapters> <icon type=chapter></icon> </tab-trigger> <tab-trigger tab=transcripts> <icon type=transcripts></icon> </tab-trigger> <tab-trigger tab=files> <icon type=download></icon> </tab-trigger> <tab-trigger tab=playlist> <icon type=playlist></icon> </tab-trigger> <tab-trigger tab=share> <icon type=share></icon> </tab-trigger> </div> <subscribe-button class="mt-1 mobile:hidden tablet:flex"></subscribe-button> </div> </div> <div class="w-full relative overflow-hidden"> <tab name=chapters> <tab-chapters></tab-chapters> </tab> <tab name=transcripts> <tab-transcripts></tab-transcripts> </tab> <tab name=files> <tab-files></tab-files> </tab> <tab name=playlist> <tab-playlist></tab-playlist> </tab> <tab name=share> <tab-share></tab-share> </tab> <tab-overflow></tab-overflow> </div> <error></error> </root> '}}]); 

plugin.php
<?php

  /**
    This is the PodlovePlugin plugin.

    This file contains the PodlovePlugin plugin. 

    @package Filmmakers4Future/PodlovePlugin
    @version 0.2
    @author  Paul-Vincent Roll <paul-vincent@filmmakersforfuture.org>
    @since   0.2
  */

  // ===== DO NOT EDIT HERE =====

  // prevent script from getting called directly
  if (!defined("URLAUBE")) { die(""); }

  class PodlovePlayerPlugin extends BaseSingleton implements Plugin {

    // CONSTANTS
    // Content 
    
    const PODLOVEWEBPLAYER = "[podloveplayer]";

    const PODLOVE_AUDIOFILES = "podlove_audioFiles";
    const PODLOVE_PODCASTTITLE = "podlove_podcastTitle";
    const PODLOVE_PODCASTLINK = "podlove_podcastLink";
    
    const PODLOVE_TITLE = "podlove_title";
    const PODLOVE_POSTER = "podlove_poster";
    const PODLOVE_DURATION = "podlove_duration";
    const PODLOVE_AUDIOSIZES = "podlove_audioSizes";
    const PODLOVE_AUDIOTITLES = "podlove_audioTitles";
    const PODLOVE_AUDIOMIME = "podlove_audioMimeTypes";
    
    const PODLOVE_DOWNLOADFILES = "podlove_downloadFiles";
    const PODLOVE_DOWNLOADSIZES = "podlove_downloadSizes";
    const PODLOVE_DOWNLOADTITLES = "podlove_downloadTitles";
    const PODLOVE_DOWNLOADMIME = "podlove_downloadMimeTypes";
    
    const PODLOVE_CHAPTERTITLES = "podlove_chapterTitles";
    const PODLOVE_CHAPTERSTARTS= "podlove_chapterStarts";
    const PODLOVE_CHAPTERLINKS = "podlove_chapterLinks";
    const PODLOVE_CHAPTERIMAGES = "podlove_chapterImages";
    
    // CONSTANTS
    // config 
    const PODLOVE_CONFIG_FEED_URL = "PODLOVE_FEED_URL";
    const PODLOVE_CONFIG_APPLE_PODCAST_ID = "PODLOVE_APPLE_PODCAST_ID";
    const PODLOVE_CONFIG_CASTBOX_ID = "PODLOVE_CASTBOX_ID";
    const PODLOVE_CONFIG_DEEZER_ID = "PODLOVE_DEEZER_ID";
    const PODLOVE_CONFIG_SOUNDCLOUD_ID = "PODLOVE_SOUNDCLOUD_ID";
    const PODLOVE_CONFIG_SPOTIFY_ID = "PODLOVE_SPOTIFY_ID";
    const PODLOVE_CONFIG_STITCHER_ID = "PODLOVE_STITCHER_ID";
    const PODLOVE_CONFIG_YOUTUBE_ID = "PODLOVE_YOUTUBE_ID";
    const PODLOVE_CONFIG_ACTIVE_TAB = "PODLOVE_ACTIVE_TAB";
    const PODLOVE_CONFIG_EMBEDDING = "PODLOVE_EMBEDDING";
    const PODLOVE_CONFIG_THEME_COLORS = "PODLOVE_THEME_COLORS";
    const PODLOVE_CONFIG_SHARE_CHANNELS = "PODLOVE_SHARE_CHANNELS";
    const PODLOVE_CONFIG_SHARE_PLAYTIME = "PODLOVE_SHARE_PLAYTIME";
    
    // VARIABLES
    
    protected static $PLAYER_BOOTSTRAP_CODE = array(); 

    // HELPER FUNCTIONS
    
    public static function bootstrapPlayer() {
      print(fhtml("<!-- Podlove Webplayer Library -->").NL);
      print(fhtml("<script src='".path2uri(__DIR__."/lib/embed.js")."'></script>").NL);
      print(fhtml("<!-- Podlove Webplayer Boostraping -->").NL);
      print(fhtml("<script>"));
      foreach (self::$PLAYER_BOOTSTRAP_CODE as $player) {
        print(fhtml($player).NL);
      }
      print(fhtml("</script>"));
      print(NL);
    }

    protected static function configure() {
      Plugins::preset(static::PODLOVE_CONFIG_FEED_URL, null);
      Plugins::preset(static::PODLOVE_CONFIG_APPLE_PODCAST_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_CASTBOX_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_DEEZER_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_SOUNDCLOUD_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_SPOTIFY_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_STITCHER_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_YOUTUBE_ID, null);
      Plugins::preset(static::PODLOVE_CONFIG_ACTIVE_TAB, null);
      
      Plugins::preset(static::PODLOVE_CONFIG_EMBEDDING, "/share.html");
      
      Plugins::preset(static::PODLOVE_CONFIG_SHARE_CHANNELS, ["twitter", "facebook", "whats-app", "mail", "link"]);
      Plugins::preset(static::PODLOVE_CONFIG_SHARE_PLAYTIME, true);
      
    }
    
    protected static function getPodlovePlayer($item) {
      // preset plugin configuration
      static::configure();
        
      $result = value($item, CONTENT);

      if (is_string($result)) {
        if (value($item, self::PODLOVE_AUDIOFILES) and value($item, self::PODLOVE_AUDIOMIME) and value($item, self::PODLOVE_AUDIOTITLES) and value($item, self::PODLOVE_AUDIOSIZES)) {
          
          # Base player configuration
          $config = [
            "version" => 5,
            "base" => path2uri(__DIR__)."/lib/",
            "theme" => [
              "tokens" => Plugins::get(static::PODLOVE_CONFIG_THEME_COLORS)
            ],
            "share" => [
              "channels" => Plugins::get(static::PODLOVE_CONFIG_SHARE_CHANNELS),
              "sharePlaytime" => Plugins::get(static::PODLOVE_CONFIG_SHARE_PLAYTIME),
              "outlet" => Plugins::get(static::PODLOVE_CONFIG_EMBEDDING)
            ]
          ];
          
          if(Plugins::get(static::PODLOVE_CONFIG_ACTIVE_TAB)){
            $config["activeTab"] = Plugins::get(static::PODLOVE_CONFIG_ACTIVE_TAB);
          }
          
          # Add subscribe button config
          if(Plugins::get(static::PODLOVE_CONFIG_FEED_URL)){
            $subscribeButton = [
              "feed" => Plugins::get(static::PODLOVE_CONFIG_FEED_URL),
              "clients" => [
                ["id" => "antenna-pod"],
                ["id" => "beyond-pod"],
                ["id" => "castro"],
                ["id" => "clementine"],
                ["id" => "downcast"],
                ["id" => "gpodder"],
                ["id" => "itunes"],
                ["id" => "i-catcher"],
                ["id" => "instacast"],
                ["id" => "overcast"],
                ["id" => "player-fm"],
                ["id" => "pocket-casts"],
                ["id" => "pocket-casts", "service" => Plugins::get(static::PODLOVE_CONFIG_FEED_URL)],
                ["id" => "google-podcasts", "service" => Plugins::get(static::PODLOVE_CONFIG_FEED_URL)],
                ["id" => "pod-grasp"],
                ["id" => "podcast-addict"],
                ["id" => "podcast-republic"],
                ["id" => "podcat"],
                ["id" => "podscout"],
                ["id" => "rss-radio"],
                ["id" => "rss"]
              ]
            ];
            if(Plugins::get(static::PODLOVE_CONFIG_APPLE_PODCAST_ID)){
              $subscribeButton["clients"][] = ["id" => "apple-podcasts", "service" => Plugins::get(static::PODLOVE_CONFIG_APPLE_PODCAST_ID)];
            }
            if(Plugins::get(static::PODLOVE_CONFIG_CASTBOX_ID)){
              $subscribeButton["clients"][] = ["id" => "castbox", "service" => Plugins::get(static::PODLOVE_CONFIG_CASTBOX_ID)];
            }
            if(Plugins::get(static::PODLOVE_CONFIG_DEEZER_ID)){
              $subscribeButton["clients"][] = ["id" => "deezer", "service" => Plugins::get(static::PODLOVE_CONFIG_DEEZER_ID)];
            } 
            if(Plugins::get(static::PODLOVE_CONFIG_SOUNDCLOUD_ID)){
              $subscribeButton["clients"][] = ["id" => "soundcloud", "service" => Plugins::get(static::PODLOVE_CONFIG_SOUNDCLOUD_ID)];
            }
            if(Plugins::get(static::PODLOVE_CONFIG_SPOTIFY_ID)){
              $subscribeButton["clients"][] = ["id" => "spotify", "service" => Plugins::get(static::PODLOVE_CONFIG_SPOTIFY_ID)];
            }
            if(Plugins::get(static::PODLOVE_CONFIG_STITCHER_ID)){
              $subscribeButton["clients"][] = ["id" => "stitcher", "service" => Plugins::get(static::PODLOVE_CONFIG_STITCHER_ID)];
            }
            if(Plugins::get(static::PODLOVE_CONFIG_YOUTUBE_ID)){
              $subscribeButton["clients"][] = ["id" => "youtube", "service" => Plugins::get(static::PODLOVE_CONFIG_YOUTUBE_ID)];
            }
            $config["subscribe-button"] = $subscribeButton;
          }
          
          
          // Base episode configuration
          $episode = [
            "version" => 5,
            "show" => [
              "title" => value($item, self::PODLOVE_PODCASTTITLE),
              "link" => value($item, self::PODLOVE_PODCASTLINK)
            ],
            "title" => value($item, self::PODLOVE_TITLE),
            "poster" => value($item, self::PODLOVE_POSTER),
            "duration" => value($item, self::PODLOVE_DURATION),
            "link" => absoluteurl(value($item, URI))
          ];
          
          // Parse audio files from markdown
          $audioFiles = explode(",", value($item, self::PODLOVE_AUDIOFILES));
          $audioSizes = explode(",", value($item, self::PODLOVE_AUDIOSIZES));
          $audioTitles = explode(",", value($item, self::PODLOVE_AUDIOTITLES));
          $audioMimeTypes = explode(",", value($item, self::PODLOVE_AUDIOMIME));
          
          // Parse download files from markdown (optional)
          $downloadFiles = explode(",", value($item, self::PODLOVE_DOWNLOADFILES));
          $downloadSizes = explode(",", value($item, self::PODLOVE_DOWNLOADSIZES));
          $downloadTitles = explode(",", value($item, self::PODLOVE_DOWNLOADTITLES));
          $downloadMimeTypes = explode(",", value($item, self::PODLOVE_DOWNLOADMIME));
          
          // Parse chapters from markdown (optional)
          $chapterTitles = explode(",", value($item, self::PODLOVE_CHAPTERTITLES));
          $chapterStarts = explode(",", value($item, self::PODLOVE_CHAPTERSTARTS));
          $chapterLinks = explode(",", value($item, self::PODLOVE_CHAPTERLINKS));
          $chapterImages = explode(",", value($item, self::PODLOVE_CHAPTERIMAGES));
          
          // Create arrays for data
          $audio = array();
          $files = array();
          $chapters = array();
          
          // Add audio files to audio array
          foreach($audioFiles as $i => $audioFile) {
            $audio[] = ["url" => $audioFile,
                        "size" => $audioSizes[$i],
                        "title" => $audioTitles[$i],
                        "mimeType" => $audioMimeTypes[$i]
                        ];
            // Also add the audio files to the download list
            $files[] = ["url" => $audioFile,
                        "size" => $audioSizes[$i],
                        "title" => "Episode: ".$audioTitles[$i],
                        "mimeType" => $audioMimeTypes[$i]
                        ];
          };
          
          // Add download files to files array
          if(value($item, self::PODLOVE_DOWNLOADFILES)){
            foreach($downloadFiles as $i => $downloadFile) {
              $files[] = ["url" => $downloadFile,
                          "size" => $downloadSizes[$i],
                          "title" => $downloadTitles[$i],
                          "mimeType" => $downloadMimeTypes[$i]
                          ];
            };
          }
          
          // Add chapters to chapter array
          if(value($item, self::PODLOVE_CHAPTERTITLES)){
            foreach($chapterTitles as $i => $chapterTitle) {
              $chapters[] = ["title" => $chapterTitle,
                          "start" => $chapterStarts[$i],
                          "href" => $chapterLinks[$i],
                          "image" => $chapterImages[$i]
                          ];
            };
          }
          
          // Add new config arrays to episode array
          $episode["audio"] = $audio;
          $episode["files"] = $files;
          $episode["chapters"] = $chapters;
                 
          # Create random id for podlove webplayer div
          $divID =  "player_".substr(uniqid(rand(),1),0,15);
          
          # Create div for podlove webplayer
          $podloveplayer = fhtml("<div id=\"%s\"></div>".NL,
                               $divID);
                    
          # Add player bootstrap code to array to add after body
          self::$PLAYER_BOOTSTRAP_CODE[] = 'window.podlovePlayer("#'.$divID.'", '.json_encode($episode, JSON_UNESCAPED_SLASHES).', '.json_encode($config, JSON_UNESCAPED_SLASHES).');';
                                
          // replace shortcode with podlove player div
          $result = str_ireplace(static::PODLOVEWEBPLAYER, $podloveplayer, $result);
        }
      }

      return $result;
    }

    // RUNTIME FUNCTIONS

    public static function plugin($content) {
      $result = $content;

      if ($result instanceof Content) {
        if ($result->isset(CONTENT)) {
          $result->set(CONTENT, static::getPodlovePlayer($result));
        }
      } else {
        if (is_array($result)) {
          // iterate through all content items
          foreach ($result as $result_item) {
            if ($result_item instanceof Content) {
              if ($result_item->isset(CONTENT)) {
                $result_item->set(CONTENT, static::getPodlovePlayer($result_item));
              }
            }
          }
        }
      }

      return $result;
    }

  }

  // register plugin
  Plugins::register(PodlovePlayerPlugin::class, "plugin", FILTER_CONTENT);
  Plugins::register(PodlovePlayerPlugin::class, "bootstrapPlayer", AFTER_BODY);